File: android/deeplinkSDK/src/main/kotlin/com/deeplink/sdk/LinkHandler.kt
package com.deeplink.sdk

import android.content.Intent
import android.net.Uri

internal class LinkHandler(private val config: DeeplinkConfig) {

    fun canHandle(uri: Uri): Boolean {
        val scheme = uri.scheme?.lowercase() ?: return false
        // Custom URI scheme (e.g. myapp://) — always a deep link
        if (scheme != "http" && scheme != "https") {
            DeeplinkLogger.log("canHandle — custom scheme: $uri")
            return true
        }
        // App Link — host must match our configured domain
        val isAppLink = uri.host == config.domain
        if (isAppLink) DeeplinkLogger.log("canHandle — App Link: $uri")
        return isAppLink
    }

    fun handle(uri: Uri): IncomingLink? {
        if (!canHandle(uri)) return null
        val segments = uri.pathSegments ?: emptyList()
        val params = uri.queryParameterNames
            .associateWith { uri.getQueryParameter(it) ?: "" }
        DeeplinkLogger.log("handle — path=${segments}, params=$params")
        return IncomingLink(url = uri, pathSegments = segments, params = params)
    }

    fun handleIntent(intent: Intent): IncomingLink? {
        val uri = intent.data ?: return null
        return handle(uri)
    }
}

File: ios/Sources/DeeplinkSDK/LinkHandler.swift
import Foundation

/// Parses and exposes an incoming deep link URL.
public struct IncomingLink {
    /// The full URL that was opened.
    public let url: URL

    /// Path components (e.g. ["product", "123"] for /product/123)
    public let pathComponents: [String]

    /// Query parameters as a dictionary.
    public let params: [String: String]

    init?(url: URL) {
        self.url = url
        self.pathComponents = url.pathComponents.filter { $0 != "/" }

        var items: [String: String] = [:]
        if let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
           let queryItems = components.queryItems {
            for item in queryItems {
                items[item.name] = item.value ?? ""
            }
        }
        self.params = items
    }
}

public final class LinkHandler {
    private let config: DeeplinkConfig

    init(config: DeeplinkConfig) {
        self.config = config
    }

    /// Returns true if this URL should be handled by the Deeplink SDK.
    ///
    /// Handles:
    /// - Universal links matching your configured domain
    /// - Custom URL scheme links (e.g. `myapp://`)
    public func canHandle(url: URL) -> Bool {
        let scheme = url.scheme?.lowercased() ?? ""
        // Custom URL scheme (not http/https) — always a deep link
        if scheme != "http" && scheme != "https" {
            DeeplinkLogger.log("Handling custom scheme URL: \(url)")
            return true
        }
        // Universal link — host must match our configured domain
        let isUniversal = url.host == config.apiBaseURL.host
        if isUniversal { DeeplinkLogger.log("Handling universal link: \(url)") }
        return isUniversal
    }

    /// Parse an incoming universal link or custom URL scheme link.
    public func handle(url: URL) -> IncomingLink? {
        guard canHandle(url: url) else { return nil }
        let link = IncomingLink(url: url)
        DeeplinkLogger.log("Parsed incoming link — path: \(link?.pathComponents ?? []), params: \(link?.params ?? [:])")
        return link
    }
}
